ouroboros-network/protocols/lib/Ouroboros/Network/Protocol/ObjectDiffusion/Type.hs

```diff
@@ -0,0 +1,298 @@
+    -- | Request a list of object identifiers from the server, and confirm a
+    -- number of outstanding object identifiers.
+    --
+    -- With 'TokBlocking' this is a blocking operation: the response will always
+    -- have at least one object identifier, and it does not expect a prompt
+    -- response: there is no timeout. This covers the case when there is nothing
+    -- else to do but wait.
+    --
+    -- With 'TokNonBlocking' this is a non-blocking operation: the response may
+    -- be an empty list and this does expect a prompt response. This covers high
+    -- throughput use cases where we wish to pipeline, by interleaving requests
+    -- for additional object identifiers with requests for objects, which
+    -- requires these requests not block.
+    --
+    -- The request gives the maximum number of object identifiers that can be
+    -- accepted in the response. This must be greater than zero in the
+    -- 'TokBlocking' case. In the 'TokNonBlocking' case either the numbers
+    -- acknowledged or the number requested __MUST__ be non-zero. In either
+    -- case, the number requested __MUST__ not put the total outstanding over
+    -- the fixed protocol limit.
+    --
+    -- The request also gives the number of outstanding object identifiers that
+    -- can now be acknowledged. The actual objects to acknowledge are known to
+    -- the server based on the FIFO order in which they were provided.
+    --
+    -- There is no choice about when to use the blocking case versus the
+    -- non-blocking case, it depends on whether there are any remaining
+    -- unacknowledged objects (after taking into account the ones acknowledged
+    -- in this message):
+    --
+    -- * The blocking case __MUST__ be used when there are zero remaining
+    --   unacknowledged objects.
+    --
+    -- * The non-blocking case __MUST__ be used when there are non-zero
+    --   remaining unacknowledged objects.
+    MsgRequestObjectIds
+      :: forall (blocking :: StBlockingStyle) objectId object.
+         SingBlockingStyle blocking
+      -> NumObjectIdsAck -- ^ Acknowledge this number of outstanding objects
+      -> NumObjectIdsReq -- ^ Request up to this number of object ids
+      -> Message (ObjectDiffusion objectId object) StIdle (StObjectIds blocking)
+    -- | Reply with a list of object identifiers for available objects, along
+    -- with the size of each object.
+    --
+    -- The list must not be longer than the maximum number requested.
+    --
+    -- In the 'StObjectIds' 'Blocking' state the list must be non-empty while in
+    -- the 'StObjectIds' 'NonBlocking' state the list may be empty.
+    --
+    -- These objects are added to the notional FIFO of outstanding object
+    -- identifiers for the protocol.
+    --
+    -- The order in which these object identifiers are returned must be the
+    -- order in which they are submitted to the mempool, to preserve dependent
+    -- objects.
+    MsgReplyObjectIds
```

Comment:
Adding a bit of spaces would be welcomed in this module.
```suggestion
    
    MsgReplyObjectIds
```

According to our style guide we use one or two empty lines (between top level definitions).